logviewer/logviewer-lib.pl

```diff
@@ -27,33 +27,135 @@ sub can_edit_log
 return 0;
 }
 
-# 
-# Returns standard 
+# get_journal_since
+# Returns a list of journalctl commands to get logs since various times,
+# which should correspond with language strings journal_since0,
+# journal_since1, journal_since2, etc.
+sub get_journal_since
+{
+return
+	("", "-f",
+	 "-b", "-S '7 days ago'", 
+	 "-S '24 hours ago'", "-S '8 hours ago'",
+	 "-S '1 hour ago'", "-S '30 minutes ago'",
+	 "-S '10 minutes ago'", "-S '3 minutes ago'",
+	 "-S '1 minute ago'");
+}
+
+# get_systemctl_cmds([force-select])
+# Returns logs for journalctl
 sub get_systemctl_cmds
 {
-my $lines = $config{'lines'} || 1000;
-return !&has_command('journalctl') ? () : (
-	{ 'cmd' => "journalctl --lines $lines -p alert..emerg",
-	  'desc' => $text{'journal_journalctl_alert_emerg'},
+my $fselect = shift;
+my $lines = $in{'lines'} ? int($in{'lines'}) : int($config{'lines'}) || 1000;
+my $journalctl_cmd = &has_command('journalctl');
+return () if (!$journalctl_cmd);
+my @rs = (
+	{ 'cmd' => "journalctl -n $lines",
+	  'desc' => $text{'journal_journalctl'},
 	  'id' => "journal-1", },
-	{ 'cmd' => "journalctl --lines $lines -p err..crit",
-	  'desc' => $text{'journal_journalctl_err_crit'},
+	{ 'cmd' => "journalctl -n $lines -x ",
+	  'desc' => $text{'journal_expla_journalctl'},
 	  'id' => "journal-2", },
-	{ 'cmd' => "journalctl --lines $lines -p notice..warning",
-	  'desc' => $text{'journal_journalctl_notice_warning'},
+	{ 'cmd' => "journalctl -n $lines -p alert..emerg",
+	  'desc' => $text{'journal_journalctl_alert_emerg'},
 	  'id' => "journal-3", },
-	{ 'cmd' => "journalctl --lines $lines -p debug..info",
-	  'desc' => $text{'journal_journalctl_debug_info'},
+	{ 'cmd' => "journalctl -n $lines -p err..crit",
+	  'desc' => $text{'journal_journalctl_err_crit'},
 	  'id' => "journal-4", },
-	{ 'cmd' => "journalctl --lines $lines -k ",
-	  'desc' => $text{'journal_journalctl_dmesg'},
+	{ 'cmd' => "journalctl -n $lines -p notice..warning",
+	  'desc' => $text{'journal_journalctl_notice_warning'},
 	  'id' => "journal-5", },
-	{ 'cmd' => "journalctl --lines $lines -x ",
-	  'desc' => $text{'journal_expla_journalctl'},
-	  'id' => "journal-6", }, 
-	{ 'cmd' => "journalctl --lines $lines",
-	  'desc' => $text{'journal_journalctl'},
+	{ 'cmd' => "journalctl -n $lines -p debug..info",
+	  'desc' => $text{'journal_journalctl_debug_info'},
+	  'id' => "journal-6", },
+	{ 'cmd' => "journalctl -n $lines -k ",
+	  'desc' => $text{'journal_journalctl_dmesg'},
 	  'id' => "journal-7", } );
+
+# Add more units from config if exists on the system
+my (%ucache, %uread);
+my $units_cache = "$module_config_directory/units.cache";
+&read_file($units_cache, \%ucache);
```

Comment:
How long is this cache kept for?

Reply:
Until module config is resaved.

Reply:
It will also clear this cache, if refresh button in the dashboard clicked. But I'm thinking of add clear button the the UI on the next PR, along side with controlling "hovering".

Reply:
Why not just keep the cache for only 1 day?

Reply:
Because in production system units are rarely changed. We could clear cache after some package was installed using Webmin.

Also, [Joe](@swelljoe) mentioned once that running `systemctl list-units --all` is slow but the fact is that running `list_systemd_services` is slow. The difference in my tests was just 100ms between reading cached file and calling command, where reading cache file was faster.

Reply:
Why not instead just have a timeout on the cache lifetime? This is what's done elsewhere in Webmin for slow operations.

Reply:
I don't find it reasonable to do it in this case.

What if we instead cleared this kind of cache on Webmin restart?

Reply:
That might not happen very often though. What's wrong with clearing based on a timeout?

Reply:
Nice, I like that more!

Reply:
> That might not happen very often though. What's wrong with clearing based on a timeout?

I don't think it's reasonable in case of _systemd_ units cache.

> Nice, I like that more!

Do you mean to clear it upon Webmin restart? If so, I will add it later to a separate PR as soon as we deal with this one.